http-client/Network/HTTP/Client/Connection.hs

```diff
@@ -184,10 +189,31 @@ openSocket tweakSocket addr =
             NS.connect sock (NS.addrAddress addr)
             return sock)
 
+-- Pick up an IP using an approximation of the happy-eyeballs algorithm:
+-- https://datatracker.ietf.org/doc/html/rfc8305
+-- 
 firstSuccessful :: [NS.AddrInfo] -> (NS.AddrInfo -> IO a) -> IO a
-firstSuccessful []     _  = error "getAddrInfo returned empty list"
-firstSuccessful (a:as) cb =
-    cb a `E.catch` \(e :: E.IOException) ->
-        case as of
-            [] -> E.throwIO e
-            _  -> firstSuccessful as cb
+firstSuccessful []        _  = error "getAddrInfo returned empty list"
+firstSuccessful addresses cb = do
+    result <- newEmptyMVar
+    either E.throwIO pure =<<
+        withAsync (tryAddresses result)
+            (\_ -> takeMVar result)
+  where
+    -- https://datatracker.ietf.org/doc/html/rfc8305#section-5
+    connectionAttemptDelay = 250 * 1000
+
+    tryAddresses result = do
+        lastException <- take 1 . filter isLeft <$> go addresses []
+        for_ lastException $ tryPutMVar result
+      where
+        go []             asyncs = mapM wait asyncs
+        go (addr : addrs) asyncs =
+            withAsync (E.evaluate =<< tryAddress addr) $ \a -> do
```

Comment:
This looks like the wrong place to do an `evaluate`. Wouldn't it be better to do this inside the `try` call?

Reply:
I added this for strictness only (to fix weird Max OS nightly tests failing), but will try to replace it with `$!` inside of the `tryAddress`.

Reply:
And the test breaks again :|
Basically, I just need to have strict behaviour of `withAsync (do { z <- tryAddress addr; pure $! z} )` that's why `evaluate`. But I'm open to any other idea.